package.json

```diff
@@ -10,7 +10,8 @@
   ],
   "dependencies": {
     "com.unity.modules.jsonserialize": "1.0.0",
-    "com.unity.editorcoroutines": "1.0.0"
+    "com.unity.editorcoroutines": "1.0.0",
+    "com.unity.nuget.newtonsoft-json": "2.0.0"
```

Comment:
We have to remove the dependency for releasing the package as verified because the verified package must not depend on non-verified ones.

Reply:
Do you mean the newtonsoft dependency has to be removed/resolved before the PR can be accepted? Because that requires the RTCParameter and RTCStats marshalling/ptr to be implemented, which is a big task.

Also, this line is wrong, as the code requires `jillejr.newtonsoft.json-for-unity": "13.0.102"` for now. The package `com.unity.nuget.newtonsoft-json": "2.0.0` crashed when parsing the RTCStats object due to IL2CPP/AOT issues.

This is why I added the `jillejr.newtonsoft.json-for-unity": "13.0.102"` as a limitation/restriction.

However, setting the package.json file requires the projects manifest.json file to be changed (scoped registry). So I'm not sure what to do now.

Reply:
Thanks, I understood.
I am investigating we can use the `com.unity.nuget.newtonsoft-json` package for our product.

Reply:
As you said, it is not clear this package supports IL2CPP or not.
https://docs.unity3d.com/Packages/com.unity.nuget.newtonsoft-json@2.0/manual/index.html

Reply:
As far as I know, newtonsoft is currently used to following ways:

`com.unity.nuget.newtonsoft-json": "2.0.0"` is needed in the NativeMethods, for example:
```
#if !UNITY_WEBGL
        [DllImport(WebRTC.Lib)]
        public static extern void PeerConnectionCreateOffer(IntPtr ptr, ref RTCOfferAnswerOptions options);
        [DllImport(WebRTC.Lib)]
        public static extern void PeerConnectionCreateAnswer(IntPtr ptr, ref RTCOfferAnswerOptions options);
#else
        [DllImport(WebRTC.Lib)]
        public static extern void PeerConnectionCreateOffer(IntPtr ptr, string options);
        [DllImport(WebRTC.Lib)]
        public static extern void PeerConnectionCreateAnswer(IntPtr ptr, string options);
#endif
```

So we need the json serialization for these kind of methods.

The third party lib `jillejr.newtonsoft.json-for-unity": "13.0.102"` was only needed for the (de)serialization of RTCStats/Parameters because `com.unity.nuget.newtonsoft-json": "2.0.0"` crashed when deserializing the RTCStatsReport object (due to IL2CPP).

I think it's possible to remove the `jillejr.newtonsoft.json-for-unity": "13.0.102"` dependency eventually, but I am not sure if we can remove `com.unity.nuget.newtonsoft-json": "2.0.0"`.

Of course, there is JsonUtility, but yea... not sure it that is the right tool for this

Reply:
原文
JsonUtility cannot serialize / deserialize only arrays.
This is the only reason I used the jillejr package.
JsonUtility will give an error if it is an array, but arrays in an object can be serialized / deserialized.
If you set the rule that the data to be marshalled must be of object type, I think it's okay to use JsonUtility.
If you run into problems with this rule, why not consider the specifications of other packages at that point and consider using JsonUtility first?

```cs
private class Data
{
    public string[] alice;
    public float[] bob;
    public string candy;
    public int dad;
}

void Start()
{
    // Array Error.
    //var arr = JsonUtility.FromJson<Data[]>("[{\"alice\":[\"fuga\", \"piyo\"], \"bob\":[1.23, 3.45, 6.78], \"candy\":\"JsonUtility\", \"dad\": 123}]");
   
    // Object Success.
    var data = JsonUtility.FromJson<Data>("{\"alice\":[\"fuga\", \"piyo\"], \"bob\":[1.23, 3.45, 6.78], \"candy\":\"JsonUtility\", \"dad\": 123}");
    Debug.Log(data.alice);
    Debug.Log(data.bob);
    Debug.Log(data.candy);
    Debug.Log(data.dad);
}
```

Reply:
If I understand it correctly, we theoretically can only use JsonUtility, with the rule that it is an Object, instead of Array. That's cool, I didn't know that.

But what do you mean with the last sentence? I think I'm confused by "consider other packages and consider JsonUtility first". I think it's fine to use other packages, and forcing everything to JsonUtility seems pretty difficult.

And just to be sure @karasusan, is there a problem for this PR that we are using a third party json lib for now? I understand that for the final version (merge to develop) it's required to use only unity packages, but removing the json lib dependency is a lot of work and would be better for another issue/PR.

Reply:
What I wanted to say is that it's a bit annoying, but I think it can be implemented with the Json Utility. Consider using other JSON libraries in future versions. about it.

Reply:
@nvanofwegen

> I understand that for the final version (merge to develop) it's required to use only unity packages

Right, this is hard but we have to do it.

> but removing the json lib dependency is a lot of work and would be better for another issue/PR.

Sure, no problem. This is what we will do.